Fix secure temp file creation

imuchnik opened the pull request:
- Bugfix Pull Request
##### COMPONENT NAME

jenkins/web_infrastructure/download_plugin
##### ANSIBLE VERSION

```
ansible 2.2.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

The execution of plugin install was failing with TypeError on file.open() and file.read() calls. tempfile.mkstemp() returns a tuple not a file path. 

No existing issue, but the bug was introduced in this PR #2887

<!--- Paste verbatim command output below, e.g. before and after your change -->

```

```

gregdek commented:
maintainer unknown ping @gregdek @resmo @robynbergeron.

gregdek commented:
Thanks @imuchnik. To the current maintainers, @jtyr please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

jtyr commented:
If this is a fix, it should fix all usages of the `mkstemp()` function. That means that the same change is needed on the line 647.

imuchnik commented:
Yes, you are correct. There are more issues with the module. I will amend the PR  as soon as I am done testing.

mikecee commented:
Line 624 in :

``` python
            self.module.atomic_move(updates_file, updates_file_orig)
```

should be updated:

``` python
            self.module.atomic_move(updates_file[1], updates_file_orig)
```

Given no usage of the handle being returned by `mkstemp` (i.e. `updates_file[0]`) wouldn't the more maintainable change have been something more like this?

``` python
     # Write the updates file
             _, updates_file = tempfile.mkstemp()
             try:
                 fd = open(updates_file, 'wb')
             except IOError:
           ...
```

or even use the handle returned by `mkstemp()`, something like:

``` python
     # Write the updates file
             update_fd, updates_file = tempfile.mkstemp()
             try:
                 fd = os.fdopen(update_fd, 'wb')   
             except IOError:
           ...
```

jtyr commented:
I like this option:

``` python
update_fd, updates_file = tempfile.mkstemp()
```

With that, we don't have to use indexes further in the code.

jtyr commented:
Hi @imuchnik. Any progress on this PR? It would be nice to have this fixed as other people having issues with this (see #3391).

brucellino commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -567,7 +567,7 @@ def _download_updates(self):
 
             try:
-                fd = open(updates_file, 'wb')
+                fd = open(updates_file[1], 'wb')
```
This change _did_ get rid of one of the errors reported in #3391, but I had to make a similar edit to the tmp file opening. When I did this, I got _different_ errors.

``` json

{"failed": true, "item": "analysis-core", "msg": "Could not replace file: (20, '/tmp/tmpd58_hC') to /var/lib/jenkins/plugins/analysis-core.jpi: [Errno 2] No such file or directory"}
```

So, this little experiment of mine didn't quite work.

imuchnik commented:
@jtyr  I had to pause on it, but will try to get it out in the next few days. There were number of other problems I had to fix as well. Sorry for the delays.

jtyr commented:
@imuchnik Do you want to finish the fix? If you don't fix it in the next 24 hours, I will go ahead and fix it myself.

imuchnik commented:
@jtyr Sorry for the delay and thank you for your patience. I added the rest. There were other issues with how Jenkins treats the presence of jpi and hpi files, so I added a small fix and filed an issue with Jenkins support. Here the issue https://issues.jenkins-ci.org/browse/JENKINS-38822 if you would like to follow the resolution.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -408,11 +408,18 @@ def _get_installed_plugins(self):
+            self.params['jenkins_home'],
+            self.params['name'])
+        if os.path.isfile(hpi_file):
+            os.rename(hpi_file, hpi_file+'.bak')
```
Please could you explain why this is here?

Edited: I see. Jenkins is using `.jpi` extension but historically there could be also plugins with extension `.hpi`. In that case I would say that we should delete the `.hpi` file **after** the new `.jpi` file was downloaded.

imuchnik replied:
@jtyr See the note I added with Jenkins issue. I left it in so there is some recourse if they want to revert.

jtyr replied:
I would say that if we managed to download the new `.jpi` file, we should be fine to delete the `.hpi`. I want to prevent the module to create clutter in the plugin directory.

imuchnik replied:
@jtyr Valid point. I updated the code.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -408,11 +408,18 @@ def _get_installed_plugins(self):
 
     def install(self):
         changed = False
+
```
This space empty line not be here.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -640,8 +647,10 @@ def _download_plugin(self, plugin_url):
         return r
 
     def _write_file(self, f, data):
+
```
This empty line should be be here.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -640,8 +647,10 @@ def _download_plugin(self, plugin_url):
         # Store the plugin into a temp file and then move it
-        tmp_f = tempfile.mkstemp()
+        tmp_f_tuple, tmp_f = tempfile.mkstemp()
+
```
This empty line should not be here.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -413,6 +413,12 @@ def install(self):
+        hpi_file = '%s/plugins/%s.hpi' % (
+            self.params['jenkins_home'],
+            self.params['name'])
+        if os.path.isfile(hpi_file):
```
There should be space in front of the `if`.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -413,6 +413,12 @@ def install(self):
+            self.params['jenkins_home'],
+            self.params['name'])
+        if os.path.isfile(hpi_file):
+            os.remove(hpi_file)
```
All this should go after the block of commented by `# Send the installation request`.

imuchnik replied:
Sorry, my bad.

jtyr commented on web_infrastructure/jenkins_plugin.py:
```diff
@@ -439,6 +439,12 @@ def install(self):
 
+                hpi_file = '%s/plugins/%s.hpi' % (
+                    self.params['jenkins_home'],
+                    self.params['name'])
```
Still missing the empty line in front of the `if` ;o)

imuchnik replied:
Sorry. Again.

jtyr commented:
@imuchnik Please could you submit your PR into the ansible/ansible repo?

abadger commented:
I believe this was submitted (and merged) at
https://github.com/ansible/ansible/pull/19096

imuchnik commented:
@abadger @jtyr That is correct. This has been migrated and merged. Thanks.
